core/src/credential.rs

```diff
@@ -0,0 +1,147 @@
+
+#![allow(unused_variables)]
+#![allow(dead_code)]
+use std::{error::Error, time::{Instant, SystemTime}};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+// cred_subject is a generic that implements trait X
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VCContext = [&'static str];
+
+pub(crate) const JSON_LD_CONTEXT:  VCContext = ["https://www.w3.org/2018/credentials/v1", "https://www.w3.org/2018/credentials/examples/v1"];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: str = "BankCard";
+
+type CredentialSubject = serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifiableCredential <'a> {
+    #[serde(rename = "@context")]
+	context:  VCContext,
+    #[serde(rename = "@id")]
+	id: String,
+    cred_type:  String,
+	issuer: String,
+	issuance_date: SystemTime,
+	subject: CredentialSubject,
+}
+
+pub struct IssuedCredential <'a> {
+    verifiable_credential:VerifiableCredential,
+    proof: *mut CredentialProof,
+}
+
+pub struct CredentialManager {
+    credential_subject: HashMap<&str, &str>,
+    context: VCContext,
+}
+
+trait VC {
```

Comment:
I don't think I quite explained myself, these trait functions should be moved to `DocumentBuilder`. As that's the public facing trait users will actually be easily able to override

Reply:
We are also missing a trait function called `getCredentialTypes`. Which returns the list of credential types (as strings) that we support

Reply:
Functions like create_credential and create_presentation we were talking about moved to `DocumentBuilder` trait indeed. VC trait and its implementation for Credential manager operates one level below. It allows you to use default `CredentialContext,` `CredTypes` and corresponding `CredentialSubjects` but gives an opportunity to overlap them by implementing your own trait, keeping the same `VerifiableCredential` struct. Unlike re-Implementing `DocumentBuilder` that let you use any type instead of `VerifiableCredential`.

Reply:
@sKudryashov When the end-user would override `DocumentBuilder` they can only override what they need. I just don't the see the value of the extra abstraction with the `CredentialManager`. If they want to do a full override they can still use the struct that we defined for that override

Reply:
My concern is that the extra abstraction seems to not add value while adding complexity. I want to make sure we take as minimal of an approach as possible since this functionality will be fairly low level. And I think having the `DocumentBuilder` be a trait that the user can leverage and optionally override to whatever measure they need is the approach that best balances ease-of-use with complexity